plonky3/src/prover/mod.rs

```diff
@@ -0,0 +1,159 @@
+            .with_witness(witness);
+
+        let publics = vec![];
+
+        let trace = circuit.generate_trace_rows();
+
+        let config = get_config(self.analyzed.degree());
+
+        let mut challenger = get_challenger();
+
+        let proof = prove(&config, &circuit, &mut challenger, trace, &publics);
+
+        let mut challenger = get_challenger();
+
+        verify(&config, &circuit, &mut challenger, &proof, &publics).unwrap();
+        Ok(serde_json::to_vec(&proof).unwrap())
+    }
+
+    pub fn verify(&self, proof: &[u8], instances: &[Vec<T>]) -> Result<(), String> {
+        let proof: Proof<_> = serde_json::from_slice(proof)
+            .map_err(|e| format!("Failed to deserialize proof: {e}"))?;
+        let publics = instances
+            .iter()
+            .flatten()
+            .map(|v| cast_to_goldilocks(*v))
+            .collect();
+
+        let config = get_config(self.analyzed.degree());
+
+        let mut challenger = get_challenger();
+
+        verify(
+            &config,
+            &PowdrCircuit::new(self.analyzed),
+            &mut challenger,
+            &proof,
+            &publics,
+        )
+        .map_err(|e| format!("Failed to verify proof: {e:?}"))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use powdr_number::GoldilocksField;
+    use powdr_pipeline::Pipeline;
+
+    use crate::Plonky3Prover;
+
+    /// Prove and verify execution
+    fn run_test_goldilocks(pil: &str) {
+        let mut pipeline = Pipeline::<GoldilocksField>::default().from_pil_string(pil.to_string());
+
+        let pil = pipeline.compute_optimized_pil().unwrap();
+        let witness_callback = pipeline.witgen_callback().unwrap();
+        let witness = pipeline.compute_witness().unwrap();
+
+        let proof = Plonky3Prover::new(&pil).prove(&witness, witness_callback);
+
+        assert!(proof.is_ok());
+    }
+
+    #[test]
+    #[should_panic = "not implemented"]
+    fn publics() {
+        let content = "namespace Global(8); pol witness x; x * (x - 1) = 0; public out = x(7);";
+        run_test_goldilocks(content);
+    }
+
+    #[test]
+    #[should_panic = "assertion failed: width >= 1"]
+    fn empty() {
+        let content = "namespace Global(8);";
+        run_test_goldilocks(content);
+    }
+
+    #[test]
+    fn add() {
+        let content = r#"
+        namespace Add(8);
+            col witness x;
+            col witness y;
+            col witness z;
+            x + y = z;
+        "#;
+        run_test_goldilocks(content);
+    }
+
+    #[test]
+    #[should_panic = "not implemented"]
+    fn challenge() {
+        let content = r#"
+        let N: int = 8;
+        namespace std::prover(N);
+            let challenge = [];
+            enum Query {
+                Hint(int)
+            }
+        
+        namespace Global(N); 
+            let beta: expr = std::prover::challenge(0, 42); 
+            col witness stage(0) x;
+            col witness stage(1) y;
+            x + beta = y + beta;
```

Comment:
```suggestion
            y = x + beta;
```
Maybe like this? Otherwise I'm afraid an enhanced optimizer would remove `beta` in the future :D